qstat.go

```diff
@@ -191,7 +191,14 @@ func buildQstatArgumentList(filters map[string]string) []string {
 func generatedQstatOputput() (string, error) {
 	//Get the medium XML file from the master repo
 
-	xmlresponse, err := http.Get("https://raw.githubusercontent.com/metrumresearchgroup/gogridengine/master/test_data/medium.xml")
+	//Default
+	xmlsource := "https://raw.githubusercontent.com/metrumresearchgroup/gogridengine/master/test_data/medium.xml"
+
+	if os.Getenv("GOGRIDENGINE_TEST_SOURCE") != "" {
+		xmlsource = os.Getenv("GOGRIDENGINE_TEST_SOURCE")
+	}
+
+	xmlresponse, err := http.Get(xmlsource)
```

Comment:
Obviously, this approach works.  BUT.  I'd like to suggest a slightly more elegant, decoupled (and therefore more testable) approach.  Rather than having your library read an environment variable, and grabbing a file off a fixed URL, instead define an XmlResource and XmlResponse type:

```
type XmlResource struct {
    data            string
}

type XmlResponse struct {
    xmlresource     XmlResource
}
```

Then, you could define a couple of simple interfaces:

```
type XmlResourceGetter interface {
    func (x *XmlResource) Get() (string, error)
}

type XmlResponseReader interface {
    func (x *XmlResponse) Read() (string, error)
}
```

Now, in your library, make your `Get()` function use `http.Get` like you already do, and in your test library, make a `mockXmlResource` that points at a file (or different URL, etc), and implements an appropriate `Get()` to satisfy the interface.  At this point, `XmlResponse` will work identically in both scenarios:

```
func (x *XmlResponse) Read() (string, error) {
    return x.xmlresource.Get()
}
```

Reply:
This entire section is going to be replaced in future versions, and this was purely a stopgap. The only reason this ability to target another source via environment variable was because the one we were using for Chokhmah exceeded the size of the channels used by prometheus, so if you tried to run tests against it without anything actually consuming data from the channels, they would just hang up.

Reply:
There are plans for a generation service for the data as well as the library to just use the test_data in the directory it's pulled down, but the long term plan of what it needs to do has to get fleshed out. We were originally just generating data, but it wasn't realistic in how it was being portrayed, so it made more sense to just collect XML and represent it. The sample we started with, however, was too large. So I just needed a temporary way to point it somewhere else so that the prometheus tests wouldn't just spin forever.

Reply:
Oh sure, whip out the "future plans" line ... j/k.  I'm not in a position to make a judgement call on this, and what you have works fine.  Certainly for the immediate needs.  It all looks good to me.

I approve this message.

Reply:
Working on updating it currently

Reply:
I implemented this style of relationships for retrieving the content, but the entire purpose of this block of code was only for testing. It's all there, however, if I need to implement a different reader to read from local file, etc

Reply:
Awesome.  In the larger arc of software engineering, this is a critical pattern to master.  It's perhaps the most important technique we have for building malleable, testable software that we can safely change over time.